orchagent/tunneldecaporch.cpp

```diff
@@ -1019,3 +1462,104 @@ bool TunnelDecapOrch::getQosMapId(const std::string &tunnelKey, const std::strin
     }
     return true;
 }
+
+void TunnelDecapOrch::updateUnhandledDecapTunnelTerms(const string &tunnel_name, const string &src_ip_str)
+{
+    SWSS_LOG_ENTER();
+    SWSS_LOG_INFO("Updating unhandled decap tunnel terms for tunnel %s with source IP %s",
+                  tunnel_name.c_str(), src_ip_str.c_str());
+
+    auto tunnel_it = unhandledDecapTerms.find(tunnel_name);
+    if (tunnel_it != unhandledDecapTerms.end())
+    {
+        for (auto term_it = tunnel_it->second.begin(); term_it != tunnel_it->second.end(); ++term_it)
+        {
+            auto &term = term_it->second;
+            if (term.term_type == TUNNEL_TERM_TYPE_P2P || term.term_type == TUNNEL_TERM_TYPE_MP2MP)
+            {
+                term.src_ip = src_ip_str;
+            }
+        }
+    }
+}
+
+void TunnelDecapOrch::processUnhandledDecapTunnelTerms(const string &tunnel_name)
+{
+    SWSS_LOG_ENTER();
+    SWSS_LOG_INFO("Processing unhandled decap tunnel terms for tunnel %s",
+                  tunnel_name.c_str());
+
+    auto tunnel_it = unhandledDecapTerms.find(tunnel_name);
+    if (tunnel_it != unhandledDecapTerms.end())
+    {
+        for (auto term_it = tunnel_it->second.begin(); term_it != tunnel_it->second.end();)
+        {
+            auto &term = term_it->second;
+            if (addDecapTunnelTermEntry(tunnel_name, term.src_ip, term.dst_ip, term.term_type, term.subnet_type))
+            {
+                term_it = tunnel_it->second.erase(term_it);
+            }
+            else
+            {
+                ++term_it;
+            }
+        }
+    }
+}
+
+inline void TunnelDecapOrch::setDecapTunnelStatus(const std::string &tunnel_name)
+{
+    auto &tunnel = tunnelTable.at(tunnel_name);
+
+    vector<FieldValueTuple> fv;
+    APPEND_IF_NOT_EMPTY(fv, tunnel, tunnel_type);
+    APPEND_IF_NOT_EMPTY(fv, tunnel, dscp_mode);
+    APPEND_IF_NOT_EMPTY(fv, tunnel, ecn_mode);
+    APPEND_IF_NOT_EMPTY(fv, tunnel, encap_ecn_mode);
+    APPEND_IF_NOT_EMPTY(fv, tunnel, ttl_mode);
+    stateTunnelDecapTable->set(tunnel_name, fv);
+}
+
+inline void TunnelDecapOrch::removeDecapTunnelStatus(const std::string &tunnel_name)
+{
+    stateTunnelDecapTable->del(tunnel_name);
+}
+
+inline void TunnelDecapOrch::setDecapTunnelTermStatus(
+    const std::string &tunnel_name, const std::string &dst_ip_str, const std::string &src_ip_str,
+    TunnelTermType term_type, const std::string &subnet_type)
+{
+    const static map<TunnelTermType, string> DecapTermTypeStrLookupTable = {
+        {TUNNEL_TERM_TYPE_P2P, "P2P"},
+        {TUNNEL_TERM_TYPE_P2MP, "P2MP"},
+        {TUNNEL_TERM_TYPE_MP2MP, "MP2MP"}};
+
+    string tunnel_term_key = tunnel_name + state_db_key_delimiter + dst_ip_str;
+    string term_type_str = DecapTermTypeStrLookupTable.at(term_type);
+    vector<FieldValueTuple> fv = {{ "term_type", term_type_str }};
+    if (term_type == TUNNEL_TERM_TYPE_P2P || term_type == TUNNEL_TERM_TYPE_MP2MP)
```

Comment:
It should be better if check `src_ip_str != """. As per SAI definition, P2MP also has src_ip.

Reply:
Yes, updated.